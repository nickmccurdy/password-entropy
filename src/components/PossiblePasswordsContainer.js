import { blue, green, red, yellow } from 'material-ui/colors'
import PropTypes from 'prop-types'
import { always, filter, findLast, identity, keys, map, pipe, sum } from 'ramda'
import React from 'react'

import PossiblePasswords from './PossiblePasswords'

export const dicewareWords = 7776

const entropyTips = [
  {
    color: red,
    minimum: 0,
    strength: 'Very Weak'
  },
  {
    color: yellow,
    minimum: 32,
    strength: 'Weak'
  },
  {
    color: blue,
    minimum: 64,
    strength: 'Strong'
  },
  {
    color: green,
    minimum: 128,
    strength: 'Very Strong'
  }
]

const toggles = {
  letters: 26,
  capitalLetters: 26,
  numbers: 10,
  symbols: 8
}

const possibleItems = {
  generic: pipe(filter(identity), keys, map(key => toggles[key]), sum),
  diceware: always(dicewareWords),
  custom: identity
}

export default function PossiblePasswordsContainer({
  length,
  options,
  optionsKey
}) {
  const possiblePasswords =
    possibleItems[optionsKey](options[optionsKey]) ** length
  const approximate = possiblePasswords > Number.MAX_SAFE_INTEGER
  const entropyBits = Math.max(0, Math.log2(possiblePasswords))
  const entropyTip = findLast(tip => entropyBits >= tip.minimum, entropyTips)

  return (
    <PossiblePasswords
      possiblePasswords={possiblePasswords}
      approximate={approximate}
      entropyBits={entropyBits}
      entropyTip={entropyTip}
    />
  )
}

PossiblePasswordsContainer.propTypes = {
  length: PropTypes.number.isRequired,
  options: PropTypes.shape({
    custom: PropTypes.number.isRequired,
    generic: PropTypes.shape({
      letters: PropTypes.bool.isRequired,
      capitalLetters: PropTypes.bool.isRequired,
      numbers: PropTypes.bool.isRequired,
      symbols: PropTypes.bool.isRequired
    }).isRequired
  }).isRequired,
  optionsKey: PropTypes.string.isRequired
}
