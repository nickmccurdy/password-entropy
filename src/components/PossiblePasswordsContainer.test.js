import { shallow } from 'enzyme'
import { mergeDeepRight } from 'ramda'
import React from 'react'

import PossiblePasswordsContainer from './PossiblePasswordsContainer'

const initialState = {
  length: 12,
  options: {
    custom: 0,
    generic: {
      letters: true,
      capitalLetters: true,
      numbers: true,
      symbols: true
    }
  },
  optionsKey: 'generic'
}

const getWrapper = (state = {}) =>
  shallow(
    <PossiblePasswordsContainer {...mergeDeepRight(initialState, state)} />
  )

describe('PossiblePasswordsContainer', () => {
  test('renders generic', () => {
    expect(getWrapper()).toMatchSnapshot()
  })

  test('renders diceware', () => {
    expect(getWrapper({ optionsKey: 'diceware' })).toMatchSnapshot()
  })

  test('renders custom', () => {
    expect(getWrapper({ optionsKey: 'custom' })).toMatchSnapshot()
  })

  test('renders custom when invalid', () => {
    expect(
      getWrapper({ options: { custom: 0 }, optionsKey: 'custom' })
    ).toMatchSnapshot()
  })

  test('renders custom with weak possiblePasswords', () => {
    expect(
      getWrapper({
        length: 32,
        options: { custom: 2 },
        optionsKey: 'custom'
      })
    ).toMatchSnapshot()
  })

  test('renders custom with strong possiblePasswords', () => {
    expect(
      getWrapper({
        length: 64,
        options: { custom: 2 },
        optionsKey: 'custom'
      })
    ).toMatchSnapshot()
  })

  test('renders custom with very strong possiblePasswords', () => {
    expect(
      getWrapper({
        length: 128,
        options: { custom: 2 },
        optionsKey: 'custom'
      })
    ).toMatchSnapshot()
  })
})
