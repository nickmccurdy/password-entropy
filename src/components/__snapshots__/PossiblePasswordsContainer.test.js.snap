// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PossiblePasswordsContainer renders custom 1`] = `
<PossiblePasswords
  approximate={false}
  entropyBits={0}
  entropyTip={
    Object {
      "color": Object {
        "100": "#ffcdd2",
        "200": "#ef9a9a",
        "300": "#e57373",
        "400": "#ef5350",
        "50": "#ffebee",
        "500": "#f44336",
        "600": "#e53935",
        "700": "#d32f2f",
        "800": "#c62828",
        "900": "#b71c1c",
        "A100": "#ff8a80",
        "A200": "#ff5252",
        "A400": "#ff1744",
        "A700": "#d50000",
      },
      "minimum": 0,
      "strength": "Very Weak",
    }
  }
  possiblePasswords={0}
/>
`;

exports[`PossiblePasswordsContainer renders custom when invalid 1`] = `
<PossiblePasswords
  approximate={false}
  entropyBits={0}
  entropyTip={
    Object {
      "color": Object {
        "100": "#ffcdd2",
        "200": "#ef9a9a",
        "300": "#e57373",
        "400": "#ef5350",
        "50": "#ffebee",
        "500": "#f44336",
        "600": "#e53935",
        "700": "#d32f2f",
        "800": "#c62828",
        "900": "#b71c1c",
        "A100": "#ff8a80",
        "A200": "#ff5252",
        "A400": "#ff1744",
        "A700": "#d50000",
      },
      "minimum": 0,
      "strength": "Very Weak",
    }
  }
  possiblePasswords={0}
/>
`;

exports[`PossiblePasswordsContainer renders custom with strong possiblePasswords 1`] = `
<PossiblePasswords
  approximate={true}
  entropyBits={64}
  entropyTip={
    Object {
      "color": Object {
        "100": "#bbdefb",
        "200": "#90caf9",
        "300": "#64b5f6",
        "400": "#42a5f5",
        "50": "#e3f2fd",
        "500": "#2196f3",
        "600": "#1e88e5",
        "700": "#1976d2",
        "800": "#1565c0",
        "900": "#0d47a1",
        "A100": "#82b1ff",
        "A200": "#448aff",
        "A400": "#2979ff",
        "A700": "#2962ff",
      },
      "minimum": 64,
      "strength": "Strong",
    }
  }
  possiblePasswords={18446744073709552000}
/>
`;

exports[`PossiblePasswordsContainer renders custom with very strong possiblePasswords 1`] = `
<PossiblePasswords
  approximate={true}
  entropyBits={128}
  entropyTip={
    Object {
      "color": Object {
        "100": "#c8e6c9",
        "200": "#a5d6a7",
        "300": "#81c784",
        "400": "#66bb6a",
        "50": "#e8f5e9",
        "500": "#4caf50",
        "600": "#43a047",
        "700": "#388e3c",
        "800": "#2e7d32",
        "900": "#1b5e20",
        "A100": "#b9f6ca",
        "A200": "#69f0ae",
        "A400": "#00e676",
        "A700": "#00c853",
      },
      "minimum": 128,
      "strength": "Very Strong",
    }
  }
  possiblePasswords={3.402823669209385e+38}
/>
`;

exports[`PossiblePasswordsContainer renders custom with weak possiblePasswords 1`] = `
<PossiblePasswords
  approximate={false}
  entropyBits={32}
  entropyTip={
    Object {
      "color": Object {
        "100": "#fff9c4",
        "200": "#fff59d",
        "300": "#fff176",
        "400": "#ffee58",
        "50": "#fffde7",
        "500": "#ffeb3b",
        "600": "#fdd835",
        "700": "#fbc02d",
        "800": "#f9a825",
        "900": "#f57f17",
        "A100": "#ffff8d",
        "A200": "#ffff00",
        "A400": "#ffea00",
        "A700": "#ffd600",
      },
      "minimum": 32,
      "strength": "Weak",
    }
  }
  possiblePasswords={4294967296}
/>
`;

exports[`PossiblePasswordsContainer renders diceware 1`] = `
<PossiblePasswords
  approximate={true}
  entropyBits={155.09775004326937}
  entropyTip={
    Object {
      "color": Object {
        "100": "#c8e6c9",
        "200": "#a5d6a7",
        "300": "#81c784",
        "400": "#66bb6a",
        "50": "#e8f5e9",
        "500": "#4caf50",
        "600": "#43a047",
        "700": "#388e3c",
        "800": "#2e7d32",
        "900": "#1b5e20",
        "A100": "#b9f6ca",
        "A200": "#69f0ae",
        "A400": "#00e676",
        "A700": "#00c853",
      },
      "minimum": 128,
      "strength": "Very Strong",
    }
  }
  possiblePasswords={4.887367798068926e+46}
/>
`;

exports[`PossiblePasswordsContainer renders generic 1`] = `
<PossiblePasswords
  approximate={true}
  entropyBits={73.55139620333959}
  entropyTip={
    Object {
      "color": Object {
        "100": "#bbdefb",
        "200": "#90caf9",
        "300": "#64b5f6",
        "400": "#42a5f5",
        "50": "#e3f2fd",
        "500": "#2196f3",
        "600": "#1e88e5",
        "700": "#1976d2",
        "800": "#1565c0",
        "900": "#0d47a1",
        "A100": "#82b1ff",
        "A200": "#448aff",
        "A400": "#2979ff",
        "A700": "#2962ff",
      },
      "minimum": 64,
      "strength": "Strong",
    }
  }
  possiblePasswords={1.3841287201e+22}
/>
`;
